### PR TITLE
Implement the Illumina refinery

### PR DESCRIPTION
**Pull request by wvauclain**
## Methods

To determine which gene to map a given probe to, we select the genes in order of priority as follows:

- Pick the gene with the most appearances in Brainarray packages for Affymetrix platforms on the same species as the input Illumina platform
- If none of the associated gene IDs appear in any Brainarray platform, an NA is emitted
- If two or more of the associated gene IDs appear an equal number of times in
  Brainarray platforms, the gene ID with the lower number is selected to break the tie.

### CODE REVIEW

File: 01_scraping_brainarray_genes/scrape_brainarray_genes.R [hunk 1]
```diff
@@ -0,0 +1,73 @@
+suppressPackageStartupMessages(library(xml2))
+suppressPackageStartupMessages(library(lazyeval))
+suppressPackageStartupMessages(library(dplyr))
+
+ensg_url <- "http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/22.0.0/ensg.asp"
```
Comment (jaclyn-taroni) [inline 01_scraping_brainarray_genes/scrape_brainarray_genes.R]: Hmm, should we make the BrainArray version a command line argument to make this easier to potentially update?

File: 01_scraping_brainarray_genes/scrape_brainarray_genes.R [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                # Rat genes are ENSRNOG
+                startsWith(Probe.Set.Name, "ENSRNOG")
+            )
+
+        mapped_gene_ids <<- c(mapped_gene_ids, unique(brainarray_df$Probe.Set.Name))
+    }))
+
+    data.frame(mapped_gene_ids) %>%
+        dplyr::count(mapped_gene_ids) %>%
+        write.table(paste0("/out/01_scraping_brainarray_genes/", species, ".tsv"), row.names=FALSE, col.names=TRUE, quote=FALSE, sep="\t")
```
Comment (jaclyn-taroni) [inline 01_scraping_brainarray_genes/scrape_brainarray_genes.R]: If you have `readr` installed already, `readr::write_tsv()` is generally more pleasant to work with!

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+out/*
+!out/02_selecting_gene_ids/
```
Comment (kurtwheeler) [inline .gitignore]: Missing newline here and in a few other files.

File: 02_selecting_gene_ids/select_gene_ids.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
+"""
+For each Illumina platform that we support:
+  - Read the dataframe of duplicate probes and the different gene IDs
+  - Read the brainarray occurrence counts for the current species
+  - Group the duplicate probes dataframe by probe ID
+  - For each group, pick the gene ID with the highest occurrence count. If there
+    are ties, then we want to pick the gene ID that is on the main Ensembl
+    assembly. If there are multiple genes on the main assembly, pick the one with
```
Comment (kurtwheeler) [inline 02_selecting_gene_ids/select_gene_ids.py]: > If there are ties, then we want to pick the gene ID that is on the main Ensembl assembly

This isn't reflected in the section of the readme that outlines the priority.

File: 02_selecting_gene_ids/select_gene_ids.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        # # having at least half the # of occurrences as the max, because there
+        # # are some occurrences like [('ENSG00000143248', 29), ('ENSG00000232995', 3)]
+        # # which definitely are not ties, but others like
+        # # [('ENSG00000230876', 11), ('ENSG00000236854', 7)] which plausibly are
+        max_num_occurrences = max(nonzero_occurrences, key=lambda x: x[1])[1]
+        # close_occurrences = [o for o in occurrences if 2 * o[1] > max_num_occurrences]
+
+        # if len(close_occurrences) == 1:
+        #     # No ties
+        #     return close_occurrences[0]
```
Comment (kurtwheeler) [inline 02_selecting_gene_ids/select_gene_ids.py]: This code can be removed?

File: 02_selecting_gene_ids/select_gene_ids.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            return None
+        elif len(nonzero_occurrences) == 1:
+            return nonzero_occurrences[0][0]
+
+        # # Now, we know that we have multiple nonzero occurrences, so we need to
+        # # check for ties. In this case, I will define a tie as two gene IDs
+        # # having at least half the # of occurrences as the max, because there
+        # # are some occurrences like [('ENSG00000143248', 29), ('ENSG00000232995', 3)]
+        # # which definitely are not ties, but others like
+        # # [('ENSG00000230876', 11), ('ENSG00000236854', 7)] which plausibly are
```
Comment (kurtwheeler) [inline 02_selecting_gene_ids/select_gene_ids.py]: This seems to be doubly commented out.

**Review by kurtwheeler — APPROVED** [review]
This looks good to me. I think I understand what the R code is doing. I can't say I'd know it if there was a better way to do something though.

File: 02_selecting_gene_ids/select_gene_ids.py (hunk 4 above)
Comment (wvauclain) [reply 02_selecting_gene_ids/select_gene_ids.py]: Good catch; this was our original decision for priority but this didn't end up actually breaking any ties and the implementation would not have been straightforward, so I removed that check. I will update this comment.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
+# Illumina Refinery
+
+A workflow for selecting which gene ID to map a given probe ID to in the case
+where a given probe maps to multiple genes.
+
+## Methods
+
+To determine which gene to map a given probe to, we select the genes in order of priority as follows:
+
+- Pick the gene with the most appearances in Brainarray packages for Affymetrix platforms on the same species as the input Illumina platform
```
Comment (jaclyn-taroni) [inline README.md]:
```suggestion
- Pick the gene with the most appearances in Brainarray packages for Affymetrix platforms of the same species as the input Illumina platform
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+- Mark Dunning, Andy Lynch and Matthew Eldridge (2021). illuminaHumanv3.db: Illumina HumanHT12v3 annotation data (chip illuminaHumanv3). R package version 1.26.0.
+- Mark Dunning, Andy Lynch and Matthew Eldridge (2021). illuminaHumanv4.db: Illumina HumanHT12v4 annotation data (chip illuminaHumanv4). R package version 1.26.0.
+- Mark Dunning, Andy Lynch and Matthew Eldridge (2021). illuminaMousev1.db: Illumina MouseWG6v1 annotation data (chip illuminaMousev1). R package version 1.26.0.
+- Mark Dunning, Andy Lynch and Matthew Eldridge (2021). illuminaMousev1p1.db: Illumina MouseWG6v1p1 annotation data (chip illuminaMousev1p1). R package version 1.26.0.
+- Mark Dunning, Andy Lynch and Matthew Eldridge (2021). illuminaMousev2.db: Illumina MouseWG6v2 annotation data (chip illuminaMousev2). R package version 1.26.0.
+- Mark Dunning, Andy Lynch and Matthew Eldridge (2021). illuminaRatv1.db: Illumina Ratv1 annotation data (chip illuminaRatv1). R package version 1.26.0.
+
+Other assets used in the computing of steo 01 come from the [Brainarray project](http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/genomic_curated_CDF.asp):
```
Comment (jaclyn-taroni) [inline README.md]:
```suggestion
Other assets used in the computing of step 01 come from the [Brainarray project](http://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/genomic_curated_CDF.asp):
```

**Review by jaclyn-taroni — APPROVED** [review]
LGTM - only some minor comments at this point

**Comment by wvauclain** [comment]
@jaclyn-taroni I have one more question before I go ahead and merge this: is
> If none of the associated gene IDs appear in any Brainarray platform, an NA is emitted

a good behavior, or should we still select one of the gene IDs even when none of them appear in Brainarray?

**Comment by jaclyn-taroni** [comment]
Hm, that seems like a case where we'd want to go to the main assembly but it sounds like that would have been quite the feat to implement? https://github.com/AlexsLemonade/illumina-refinery/pull/1#discussion_r675033383

**Comment by wvauclain** [comment]
I can do some more digging, but from my initial impressions, the frontend says whether or not something is in the main assembly but I could not find anywhere on the API that indicates it.

**Comment by jaclyn-taroni** [comment]
Gotcha - if that holds true, we should probably pick the lowest ID like we do to break ties.

**Comment by wvauclain** [comment]
@jaclyn-taroni I dug through Ensembl's website's Perl code and figured out how they detect whether or not a gene is on the main assembly. It turns out that if we combine a couple of the API endpoints, we can approximate the same check, so I added that to the list of tiebreakers.

File: 02_selecting_gene_ids/select_gene_ids.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            if len(max_occurrences) == 1:
+                return max_occurrences[0][0]
+
+            # This filter failed, let's try another one
+            genes = [o[0] for o in max_occurrences]
+        else:
+            # If everything occurs 0 times, we will need to try another filter method
+            genes = [o[0] for o in occurrences]
+
+        # Now we have a tie. We need to gather more information on the genes in order ot progress.
```
Comment (kurtwheeler) [inline 02_selecting_gene_ids/select_gene_ids.py]:
```suggestion
        # Now we have a tie. We need to gather more information on the genes in order to progress.
```

**Comment by jaclyn-taroni** [comment]
Spot-checked a few of the human Ensembl gene IDs that were added in https://github.com/AlexsLemonade/illumina-refinery/pull/1/commits/00d567f4e4c344fe6dcd6d3426849d6a32a3ab6c on Ensembl's website and it looks as I would expect based on what you've included! 

Initially, it was less obvious to me (from looking at the website alone) that what we're retrieving for mouse and rat is equivalent. ([I think we'd expect only mouse and human to have this main (or primary) assembly, alternative distinction.](https://www.ensembl.org/info/genome/genebuild/haplotypes_patches.html
)). But it looks like the final "take the lower Ensembl ID" step is working as expected in the mouse and rat cases that initially gave me pause.

**Comment by wvauclain** [comment]
@jaclyn-taroni I re-ran the mouse and rat platforms with some debugging print statements, and it looks like none of the mouse or rat genes come from alternate sequences.

The system does correctly detect that the tied gene IDs are both on the main platform, though, and applies the appropriate tiebreaker.

**Review by jaclyn-taroni — APPROVED** [review]

**Pull request merged**